plugins/woocommerce/tests/legacy/bootstrap.php

```diff
@@ -172,26 +180,33 @@ private function initialize_hpos() {
 	 * Re-initialize the dependency injection engine.
 	 *
 	 * The dependency injection engine has been already initialized as part of the Woo initialization, but we need
-	 * to replace the registered read-only container with a fully configurable one for testing.
+	 * to replace the registered runtime container with one with extra capabilities for testing.
 	 * To this end we hack a bit and use reflection to grab the underlying container that the read-only one stores
 	 * in a private property.
 	 *
-	 * Additionally, we replace the legacy/function proxies with mockable versions to easily replace anything
-	 * in tests as appropriate.
+	 * Note also that TestingContainer replaces the instance of LegacyProxy with an instance of MockableLegacyProxy.
+	 *
+	 * @param bool $use_old_container The underlying container is the old ExtendedContainer class. This parameter will disappear in WooCommerce 10.0.
 	 *
 	 * @throws \Exception The Container class doesn't have a 'container' property.
 	 */
-	private function initialize_dependency_injection() {
+	private function initialize_dependency_injection( bool $use_old_container ) {
 		try {
 			$inner_container_property = new \ReflectionProperty( \Automattic\WooCommerce\Container::class, 'container' );
 		} catch ( ReflectionException $ex ) {
 			throw new \Exception( "Error when trying to get the private 'container' property from the " . \Automattic\WooCommerce\Container::class . ' class using reflection during unit testing bootstrap, has the property been removed or renamed?' );
 		}
 
 		$inner_container_property->setAccessible( true );
-		$inner_container = $inner_container_property->getValue( wc_get_container() );
 
-		$inner_container->replace( LegacyProxy::class, MockableLegacyProxy::class );
+		$container       = wc_get_container();
+		$inner_container = $inner_container_property->getValue( $container );
+		if ( $use_old_container ) {
+			$inner_container->replace( LegacyProxy::class, MockableLegacyProxy::class );
```

Comment:
I don't think this is a blocker, but when I run tests with the `USE_OLD_DI_CONTAINER` the system fails on my end.

Although the `$use_old_container` is set correctly, the inner container appears to be instantiated as RuntimeContainer. Maybe a race condition?


The command:
`USE_OLD_DI_CONTAINER=YES pnpm test:php`

The error:
```
Error in bootstrap script: Error:
Call to undefined method Automattic\WooCommerce\Internal\DependencyManagement\RuntimeContainer::replace()
#0 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(96): WC_Unit_Tests_Bootstrap->initialize_dependency_injection(true)
#1 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(329): WC_Unit_Tests_Bootstrap->__construct()
#2 /Users/chris/Repos/woocommerce/plugins/woocommerce/tests/legacy/bootstrap.php(336): WC_Unit_Tests_Bootstrap::instance()
#3 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/Util/FileLoader.php(66): include_once('/Users/chris/Re...')
#4 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/Util/FileLoader.php(49): PHPUnit\Util\FileLoader::load('/Users/chris/Re...')
#5 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(567): PHPUnit\Util\FileLoader::checkAndLoad('./tests/legacy/...')
#6 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(347): PHPUnit\TextUI\Command->handleBootstrap('./tests/legacy/...')
#7 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(114): PHPUnit\TextUI\Command->handleArguments(Array)
#8 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/src/TextUI/Command.php(99): PHPUnit\TextUI\Command->run(Array, true)
#9 phpvfscomposer:///Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/phpunit(106): PHPUnit\TextUI\Command::main()
#10 /Users/chris/Repos/woocommerce/plugins/woocommerce/vendor/bin/phpunit(118): include('phpvfscomposer:...')
```

Reply:
I can't reproduce this but it looks like a race condition indeed. In `WC_Unit_Tests_Bootstrap::__construct` the existence of the `USE_OLD_DI_CONTAINER` variable is checked, if found then the `WOOCOMMERCE_USE_OLD_DI_CONTAINER` is defined; looks like by then the main container (`src/Container.php`) has been instantiated already and thus it's using a `RuntimeContainer`. But I don't understand why this is not happening for me (and others who have already tested). Is there anything "unusual" in your setup?

Reply:
Thanks for looking into this @Konamiman!

> Is there anything "unusual" in your setup?

I realized that I had hardcoded the `should_use` value during testing. After reverting this change, everything works as expected. 😅

Reply:
Well that was unusual, for sure 😆 (I can't laugh too hard, though, as I have done similar things in the past)